File: platform/src/main/java/ru/abondin/hreasy/platform/service/EmployeeService.java
package ru.abondin.hreasy.platform.service;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Sort;
import org.springframework.data.relational.core.query.Criteria;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.client.HttpClientErrorException;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import ru.abondin.hreasy.platform.auth.AuthContext;
import ru.abondin.hreasy.platform.repo.employee.EmployeeRepo;
import ru.abondin.hreasy.platform.service.currentproject.EmployeeProjectSecurityValidator;
import ru.abondin.hreasy.platform.service.dto.CurrentProjectRole;
import ru.abondin.hreasy.platform.service.dto.EmployeeDto;
import ru.abondin.hreasy.platform.service.dto.EmployeeProjectChangesDto;
import ru.abondin.hreasy.platform.service.mapper.EmployeeDtoMapper;

@Service
@Slf4j
@RequiredArgsConstructor
public class EmployeeService {
    private final EmployeeDtoMapper mapper;
    private final DateTimeService dateTimeService;
    private final EmployeeRepo emplRepo;
    private final EmployeeProjectSecurityValidator employeeProjectSecurityValidator;

    public Flux<EmployeeDto> findAll(AuthContext auth, boolean includeFired) {
        log.debug("Find all employees from {} account", auth.getEmail());
        Criteria criteria = null;
        if (!includeFired) {
            criteria = addNotFiredCriteria(criteria);
        }
        return emplRepo.findDetailed(criteria,
                        Sort.by("display_name")
                ).map(e -> mapper.employeeWithSkills(e, auth.getEmployeeInfo().getEmployeeId()))
                .doOnNext(empl -> employeeProjectSecurityValidator.setToNullUngrantedFields(empl, auth));
    }


    public Mono<EmployeeDto> find(int employeeId, AuthContext auth) {
        log.debug("Find {} employee from {} account", employeeId, auth.getEmail());
        return emplRepo.findDetailed(employeeId)
                .map(e -> mapper.employeeWithSkills(e, auth.getEmployeeInfo().getEmployeeId()))
                .switchIfEmpty(Mono.error(new HttpClientErrorException(HttpStatus.NOT_FOUND,
                        "Employee with ID=" + employeeId + " not found")))
                .doOnNext(empl -> employeeProjectSecurityValidator.setToNullUngrantedFields(empl, auth));
    }


    private Criteria addNotFiredCriteria(Criteria criteria) {
        var dateOfDismissalColumn = "date_of_dismissal";
        var notFiredCriteria = criteria == null ? Criteria.where(dateOfDismissalColumn)
                : criteria.and(dateOfDismissalColumn);
        return notFiredCriteria.isNull().or(dateOfDismissalColumn).greaterThan(dateTimeService.now());

    }

    public Flux<CurrentProjectRole> currentUserRoles(AuthContext auth) {
        log.debug("Find all current user roles by {}", auth.getUsername());
        return emplRepo.uniqueCurrentProjectRoles(dateTimeService.now()).map(CurrentProjectRole::new);
    }

    public Flux<EmployeeProjectChangesDto> employeeProjectChanges(AuthContext auth, int employeeId) {
        log.debug("Find project changes for employee {} by ", employeeId, auth.getUsername());
        return emplRepo.employeeProjectChanges(employeeId).map(mapper::employeeProjectChangesFromEntry);
    }
}

File: platform/src/main/kotlin/ru/abondin/hreasy/platform/service/EmployeeService.kt
package ru.abondin.hreasy.platform.service

import org.springframework.data.domain.Sort
import org.springframework.data.r2dbc.query.Criteria
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Component
import org.springframework.web.client.HttpClientErrorException
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import ru.abondin.hreasy.platform.BusinessError
import ru.abondin.hreasy.platform.api.employee.EmployeeDto
import ru.abondin.hreasy.platform.api.employee.SimpleDictDto
import ru.abondin.hreasy.platform.config.AuthContext
import ru.abondin.hreasy.platform.logger
import ru.abondin.hreasy.platform.repo.EmployeeDetailedEntry
import ru.abondin.hreasy.platform.repo.EmployeeEntry
import ru.abondin.hreasy.platform.repo.EmployeeRepo

fun employeeDisplayName(lastname: String?, firstname: String?, patronymicName: String?) = listOfNotNull(lastname, firstname, patronymicName).joinToString(" ");

@Component
class EmployeeService(
        val emplRepo: EmployeeRepo,
        val dateTimeService: DateTimeService) {
    val logger = logger();

    fun findAll(auth: AuthContext, includeFired: Boolean = false): Flux<EmployeeDto> {
        logger.debug("Find all employees from ${auth.email} account");
        //TODO Add filtering and ordering
        var criteria: Criteria? = null;
        criteria = if (includeFired) criteria else addNotFiredCriteria(criteria);
        return emplRepo.findDetailed(criteria,
                Sort.by("lastname", "firstname", "patronymicName")
        ).map { e -> employeeEntryToDtoMap(e) };
    }

    fun find(employeeId: Int, auth: AuthContext): Mono<EmployeeDto> {
        logger.debug("Find {{employeeId}} employee from ${auth.email} account");
        return emplRepo.findDetailed(employeeId)
                .map { e -> employeeEntryToDtoMap(e) }
                .switchIfEmpty(Mono.error(HttpClientErrorException(HttpStatus.NOT_FOUND, "Employee with ID=${employeeId} not found")));
    }

    fun updateCurrentProject(employeeId: Int, newCurrentProjectId: Int, auth: AuthContext): Mono<Boolean> {
        logger.info("Update current project ${newCurrentProjectId} for employee ${employeeId}" +
                "by ${auth.email}");
        return emplRepo.updateCurrentProject(employeeId, newCurrentProjectId).map { updatedRowsCount -> updatedRowsCount > 0 };
    }

    fun updateCurrentProject(newCurrentProjectId: Int, auth: AuthContext): Mono<Boolean> {
        if (auth.employeeInfo == null) {
            throw BusinessError("errors.no.employee.for.auth", arrayOf(auth.email));
        }
        logger.info("Update own current project ${newCurrentProjectId} - ${auth.email} (${auth.employeeInfo.id})");
        return emplRepo.updateCurrentProject(auth.employeeInfo.id, newCurrentProjectId).map { updatedRowsCount -> updatedRowsCount > 0 };
    }

    private fun addNotFiredCriteria(criteria: Criteria?): Criteria {
        val notFiredCriteria =
                criteria?.and("date_of_dismissal") ?: Criteria.where("date_of_dismissal");

        return notFiredCriteria.isNull().or("date_of_dismissal").greaterThan(dateTimeService.now());

    }


}

/**
 * Map EmployeeEntry or EmployeeEntryDetailed to EmployeeDto
 */
fun employeeEntryToDtoMap(entry: EmployeeEntry): EmployeeDto {
    return EmployeeDto(
            id = entry.id,
            lastname = entry.lastname,
            firstname = entry.firstname,
            patronymicName = entry.patronymicName,
            displayName = employeeDisplayName(entry.lastname, entry.firstname, entry.patronymicName),
            birthday = entry.birthday,
            sex = entry.sex,
            department = nullableDictFromNullableValues(entry.departmentId, (entry as? EmployeeDetailedEntry)?.departmentName),
            currentProject = nullableDictFromNullableValues(entry.currentProjectId, (entry as? EmployeeDetailedEntry)?.currentProjectName),
            email = entry.email,
            phone = entry.phone,
            skype = entry.skype
    );
}

fun nullableDictFromNullableValues(id: Int?, name: String?): SimpleDictDto? = if (id == null) null else SimpleDictDto(id!!
        , name);